Add onboarding to monitoring (#1052)
redhat-appstudio/infra-deployments: components/monitoring/README.md

> ###On boarding to services to Monitoring

Reviewer: The conversion to Markdown didn't quite go as expected: https://www.markdownguide.org/basic-syntax/#headings

The headers need a space after the #, and the top level H1 is a single # while the H2 would be a ##.

#1. Metrics exporter services

> a. The intended service should export the metrics from the application so that prometheus is able to understand it. 

Reviewer: This list got all run together into one line, which makes it hard to read, so use the format in https://www.markdownguide.org/basic-syntax/#lists-1  instead of a. b. c.

> b. For refernce https://prometheus.io/docs/instrumenting/writing_exporters/

Reviewer: ```suggestion
b. For reference, see https://prometheus.io/docs/instrumenting/writing_exporters/
```

Author: done

c. Exported port, service, route should be accessible to prometheus service.
[Here](https://github.com/redhat-appstudio/service-provider-integration-operator/blob/main/config/rbac/auth_proxy_service.yaml) is example for the spi-system
```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    control-plane: controller-manager
  name: controller-manager-metrics-service
  namespace: system
spec:
  ports:
  - name: metrics
    port: 8443
    protocol: TCP
    targetPort: https
  selector:
    control-plane: controller-manager
    app.kubernetes.io/name: service-provider-integration-operator
```

#2. Service monitors
a.
i. If service monitor is for prometheus it self

> * Add service monitor declaration for scrapping

Reviewer: This is rendering strangely because the Markdown is not correct, but I think you could combine these two lines (29 and 30) and change scrapping -> scraping.

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  namespace: appstudio-workload-monitoring
  name: prometheus
  labels:
    prometheus: appstudio-workload
spec:
  endpoints:
  - bearerTokenSecret:
      key: ""
    interval: 15s
    path: /metrics
    port: oauth2-proxy
    scheme: HTTPS
    tlsConfig:
      caFile: /etc/prometheus/tls/tls.crt # volume mounted in the `prometheus` container of the prometheus pods
      serverName: "prometheus-oauth2.appstudio-workload-monitoring.svc"
  namespaceSelector:
    matchNames:
    - appstudio-workload-monitoring
  selector:
    matchLabels:
      app.kubernetes.io/instance: monitoring-workload-in-cluster
```
* [Here](https://github.com/redhat-appstudio/infra-deployments/blob/main/components/monitoring/prometheus/base/prometheus-servicemonitors.yaml) is an example servicemonitor for prometheus itself
ii. If the service is for getting other components added to prometheus monitoring

> * Add a Sevice monitoring decalration for scrapping

Reviewer: This is rendering strangely because the Markdown is not correct, but I think you could combine these two lines (58 and 59) and change scrapping -> scraping, plus decalration -> declaration.

Author: I want to make it as declaration of servicemonitor for scarping.

      - Add servicemonitor declaration for scrapping the intended service
      - [Here](https://github.com/redhat-appstudio/infra-deployments/blob/main/components/monitoring/prometheus/base/prometheus-servicemonitors.yaml) is an example servicemonitor for prometheus itself
      
      ```yaml
      apiVersion: monitoring.coreos.com/v1
      kind: ServiceMonitor
      metadata:
        namespace: appstudio-workload-monitoring
        name: prometheus
        labels:
          prometheus: appstudio-workload
      spec:
        endpoints:
        - bearerTokenSecret:
            key: ""
          interval: 15s
          path: /metrics
          port: oauth2-proxy
          scheme: HTTPS
          tlsConfig:
            caFile: /etc/prometheus/tls/tls.crt 
            serverName: "prometheus-oauth2.appstudio-workload-monitoring.svc"
        namespaceSelector:
          matchNames:
          - appstudio-workload-monitoring
        selector:
          matchLabels:
            app.kubernetes.io/instance: monitoring-workload-in-cluster
      ```
     
      

  - If the servicemonitor is for getting other components added to prometheus monitoring
      

>       - Add a sevicemonitor decalration for scrapping the intended service

Reviewer: ```suggestion
      - Add a sevicemonitor declaration for scraping the intended service
```

      
          ```yaml
          apiVersion: monitoring.coreos.com/v1
          kind: ServiceMonitor
          metadata:
            labels:
              prometheus: appstudio-workload <label getting discovered by prometheus-operator>
            name: release-service

>             namespace: release-service `<name of the namespace the service is in>`

Reviewer: ```suggestion
            namespace: release-service <name of the namespace the service is in>
```

          spec:
            endpoints:
            - path: /metrics
              port: https
              scheme: https
              bearerTokenSecret:

>                 name: `<secret for accessing the endpoint path on the service>`

Reviewer: ```suggestion
                name: <secret for accessing the endpoint path on the service>
```

                key: token
              tlsConfig:
                insecureSkipVerify: true
            namespaceSelector:
              matchNames:
              - release-service
            selector:
              matchLabels:
                control-plane: controller-manager
          ```

>       - Note: `make sure the namespace` is the namespace for the intended service itself, in this case `release-service`.

Reviewer: ```suggestion
      - Note: The namespace of the ServiceMonitor matches the namespace for the service we are scraping, in this case, `release-service`.
```

    b. It should have the accessible port and route to the service (or service url)

    c. Access token or service accounts as required.

    d. Use label selectors to select the desired service uniquely in the cluster.

### 3. View access to the exporter service for Prometheus

    a. Prometheus should have view access to the metrics exporter service namespace

    b. Add the Rolebinding to give prometheus view access. 
  - [Here](https://github.com/redhat-appstudio/infra-deployments/blob/main/components/monitoring/prometheus/base/prometheus-view.yaml) is an example providing prometheus view access to the cluster

  ```yaml
  kind: ClusterRoleBinding
  apiVersion: rbac.authorization.k8s.io/v1
  metadata:
    name: prometheus-view
  subjects:
  - kind: ServiceAccount
    name: prometheus-k8s
    namespace: appstudio-workload-monitoring
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: view
  ```

#### 4. Grafana dashboards - manual export

>     a. Create a new folder in grafana for your service. (In the left nav, + Create Folder)

Reviewer 2: Which grafana? Where?

Author: added link there for the stage deployment.

>     b. Create a dashboard for your team's view on your service's service level indicators. (After navigating to your folder, + Create Dashboard)

Reviewer: ```suggestion
    b. [Create a dashboard](https://grafana.com/docs/grafana/v9.0/dashboards/) for your team's view of your service's Service Level Indicators. (After navigating to your folder, + Create Dashboard)
```

    c. Add tiles on the dashboard to track your initial set of service level indicators. If you correctly added your servicemonitor to the stage Prometheus datasource, it will show up in the Query list when you edit a tile.

    d. Export the dashboard definition in JSON format. (At the top of the screen, the icon with 3 dots lets you "Share dashboard or panel". Select Export... Save to file.)

    e.  Store the dashboard definition in git, in infra-deployments. 

>   - [Here](https://github.com/redhat-appstudio/infra-deployments/blob/main/components/monitoring/grafana/base/dashboards/example.json) is an example for the default dashboard.

Reviewer: ```suggestion
  - [Here](https://github.com/redhat-appstudio/infra-deployments/blob/main/components/monitoring/grafana/base/dashboards/example.json) is an example for the default dashboard.
  
     f. Add your dashboard to the [kustomization file](https://github.com/redhat-appstudio/infra-deployments/blob/9476ade147761cec39239be4f92639b2518c8bb7/components/monitoring/grafana/base/kustomization.yaml#L15) to automatically add it to future deployments.
```
